### PR TITLE
Bug 1902059: bootstrap: always wire bound SA keys

### PR DESCRIPTION
**Pull request by stlaz**
/assign @tnozicka
cc @marun 

depends on https://github.com/openshift/installer/pull/4469

### CODE REVIEW

**Comment by stlaz** [comment]
@marun @tnozicka @abhinavdahiya I updated the PR. The lack of "Validate()" unit testing is concerning.

In my ideal world, the signing key files would have been provided by flags and validated as such.

File: pkg/cmd/render/render_test.go [hunk 1]
```diff
@@ -408,75 +408,69 @@ spec:
...
-				issuerKeyFile := cfg.APIServerArguments["service-account-signing-key-file"]
-				// FIXME: Update when we wire a real default key
-				expectedIssuerKeyFile := kubecontrolplanev1.Arguments{"/etc/kubernetes/secrets/service-account.key"}
-				if !reflect.DeepEqual(issuerKeyFile, expectedIssuerKeyFile) {
-					return fmt.Errorf("expected the service-account-signing-key-file to be %q, but it was %q", expectedIssuerKeyFile, issuerKeyFile)
-				}
-				return nil
-			},
 		},
+		//
```
Comment (marun) [inline pkg/cmd/render/render_test.go:419]: Merging commented-out code seems like an anti-pattern to me, are you intending to address before merge or track resolution with bz etc?

**Comment by marun** [comment]
> @marun @tnozicka @abhinavdahiya I updated the PR. The lack of "Validate()" unit testing is concerning.
> 
> In my ideal world, the signing key files would have been provided by flags and validated as such.

The installer is validating the keys iirc. Presumably to ensure fail-fast well before rendering?

File: pkg/cmd/render/render.go [hunk 2]
```diff
@@ -180,11 +186,30 @@ func (r *renderOpts) Run() error {
...
+		}
+
+		// the private key is missing => generate the keypair
+		pubPEM, privPEM, err := generateKeyPairPEM()
+		if err != nil {
+			return fmt.Errorf("failed to generate an RSA private key for bound SA tokens singing: %v", err)
+		}
+
+		if err := ioutil.WriteFile(boundSAPrivatePath, privPEM, os.FileMode(0600)); err != nil {
+			return fmt.Errorf("failed to write private key for bound SA tokens signing: %v", err)
```
Comment (marun) [inline pkg/cmd/render/render.go:205]: nit: s/tokens/token/

File: pkg/cmd/render/render.go [hunk 3]
```diff
@@ -180,11 +186,30 @@ func (r *renderOpts) Run() error {
...
+		}
+
+		if err := ioutil.WriteFile(boundSAPublicPath, pubPEM, os.FileMode(0644)); err != nil {
+			return fmt.Errorf("failed to write public key for bound SA tokens verification: %v", err)
```
Comment (marun) [inline pkg/cmd/render/render.go:209]: nit: s/tokens/token/

File: pkg/cmd/render/render.go [hunk 4]
```diff
@@ -180,11 +186,30 @@ func (r *renderOpts) Run() error {
...
 		}
 	}
-	if _, err := os.Stat(filepath.Join(r.generic.AssetInputDir, "bound-service-account-signing-key.key")); err == nil {
-		if _, err := os.Stat(filepath.Join(r.generic.AssetInputDir, "bound-service-account-signing-key.pub")); err == nil {
-			renderConfig.UserProvidedBoundSASigningKey = true
+
+	boundSAPublicPath := filepath.Join(r.generic.AssetInputDir, "bound-service-account-signing-key.pub")
+	boundSAPrivatePath := filepath.Join(r.generic.AssetInputDir, "bound-service-account-signing-key.key")
+	_, privStatErr := os.Stat(boundSAPrivatePath)
+	if privStatErr != nil {
```
Comment (marun) [inline pkg/cmd/render/render.go:193]: nit: Maybe skip a level of indentation?

```golang
if os.IsNotExist(...) {
  // create and write key
} else if err != nil {
  // return error
}

File: pkg/cmd/render/render.go [hunk 5]
```diff
@@ -107,6 +111,10 @@ func (r *renderOpts) Validate() error {
 		return errors.New("missing etcd serving CA: --manifest-etcd-serving-ca")
 	}
 
+	if err := validateBoundSATokensSigningKeys(r.generic.AssetInputDir); err != nil {
```
Comment (marun) [inline pkg/cmd/render/render.go:114]: Maybe set a render opt here indicating whether generation is required?

```golang
ok, err := validateBoundSATokensSigningKeys(...)
if err != nil {
  return err
}
r.BoundTokenSigningKeysProvided = ok
```

That would ensure that validation of a provided keypair could be limited to the `Validate` function, and `Run` could just respond to the option.

File: pkg/cmd/render/render_test.go [hunk 6]
```diff
@@ -218,7 +218,7 @@ func TestRenderCommand(t *testing.T) {
 		testFunction  func(cfg *kubecontrolplanev1.KubeAPIServerConfig) error
 	}{
 		{
-			name: "scenario 1 checks feature gates",
+			name: "checks feature gates",
```
Comment (marun) [inline pkg/cmd/render/render_test.go:221]: For my info, why is it desirable to remove scenario numbering? This seems like an unrelated non-functional change.

File: pkg/cmd/render/render_test.go [hunk 7]
```diff
@@ -408,75 +408,69 @@ spec:
 			},
 		},
 		{
-			name: "scenario 8 no user provided bound-sa-signing-keys",
+			name: "no user provided bound-sa-signing-keys", // makes sure there are no errors generating the keys
```
Comment (marun) [inline pkg/cmd/render/render_test.go:411]: nit: Maybe update the name of the test to reflect the intent rather than adding a comment?

File: pkg/cmd/render/render_test.go (hunk 6 above)
Comment (stlaz) [reply pkg/cmd/render/render_test.go:221]: You are getting the name of the test in a fail report anyway, so you have something to look up the test by. If you remove a test somewhere in the middle, or it feels only right to add a test in the middle of the array of the tests, you would have to to renumber. I do not see the value of test numbers.

File: pkg/cmd/render/render.go [hunk 8]
```diff
@@ -180,11 +186,30 @@ func (r *renderOpts) Run() error {
...
+	_, privStatErr := os.Stat(boundSAPrivatePath)
+	if privStatErr != nil {
+		if !os.IsNotExist(privStatErr) {
+			return fmt.Errorf("failed to access %s: %v", boundSAPrivatePath, privStatErr)
+		}
+
+		// the private key is missing => generate the keypair
+		pubPEM, privPEM, err := generateKeyPairPEM()
+		if err != nil {
+			return fmt.Errorf("failed to generate an RSA private key for bound SA tokens singing: %v", err)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/cmd/render/render.go:201]: > s/singing/signing/
I might have gotten a bit too festive back then :thinking:

File: pkg/cmd/render/render_test.go (hunk 1 above)
Comment (stlaz) [reply pkg/cmd/render/render_test.go:419]: I think I wanted to make this mergable before the Holidays, but since I failed, I might just give it a try and address it before merge.

File: pkg/cmd/render/render.go (hunk 4 above)
Comment (stlaz) [reply pkg/cmd/render/render.go:193]: I like the "something happened, what is it" approach maybe a little bit better.

File: pkg/cmd/render/render.go (hunk 5 above)
Comment (stlaz) [reply pkg/cmd/render/render.go:114]: passing validation does not mean the keys were provided

Comment (stlaz) [reply pkg/cmd/render/render.go:114]: but I understand your point, I think I dropped the idea of validation setting the option because it should merely validate and not provide further guidance of what to do with the provided options/files
